utils/gen-snapshot-tests.js

```diff
@@ -95,41 +96,100 @@ function generateTestFile(root, examples, componentFileMap) {
 }
 
-for (const section of styleguideConfig.sections) {
-    const content = fs.readFileSync(section.content, "utf8");
+function readExamplesFromDocument(documentPath) {
+    if (!fs.existsSync(documentPath)) {
+        return [];
+    }
+    const content = fs.readFileSync(documentPath, "utf8");
     const tokens = marked.lexer(content);
     const examples = tokens
         .filter((token) => token.type === "code")
         .map((token) => token.text);
-    if (examples.length === 0) {
-        // eslint-disable-next-line no-console
-        console.warn(`no examples for section ${section.name}`);
-        continue;
+
+    return examples;
+}
+
+function extractExamplesAndComponentsForFile(file, collation) {
+    const src = fs.readFileSync(file, "utf8");
+    const match =
+        src.match(/export default class \s*(\w+)/) ||
+        src.match(/export default \s*\w+\(\s*(\w+)/);
+
+    const alreadyMapped = collation.componentFileMap[match[1]] === file;
+    if (!alreadyMapped) {
+        // Only gather examples for this component file if
+        // we didn't see it already.
+        collation.componentFileMap[match[1]] = file;
+
+        const componentDoc = path.join(
+            path.dirname(file),
+            path.basename(file, path.extname(file)),
+        ) + ".md";
+
+        collation.examples = collation.examples.concat(
+            readExamplesFromDocument(componentDoc),
+        );
+    }
+}
+
+function extractExamplesAndComponentFiles(section, collation) {
+    // Add any examples fromt he section content.
+    collation.examples = collation.examples.concat(
+        readExamplesFromDocument(section.content),
+    );
+
+    // Then process any component files and add their stuff too.
+    if (section.components) {
+        const componentGlobs = typeof section.components === "string"
+            ? [section.components]
+            : section.components;
+        for (const componentGlob of componentGlobs) {
+            const files = glob.sync(componentGlob, {ignore: ignored});
+            for (const file of files) {
+                extractExamplesAndComponentsForFile(file, collation);
+            }
+        }
     }
 
-    if (section.content) {
-        const root = path.dirname(section.content);
-
-        if (section.components) {
-            glob(section.components, {ignore: ignored}, (err, files) => {
-                const componentFileMap = {};
-                for (const file of files) {
-                    const src = fs.readFileSync(file, "utf8");
-                    const match =
-                        src.match(/export default class \s*(\w+)/) ||
-                        src.match(/export default \s*\w+\(\s*(\w+)/);
-                    componentFileMap[match[1]] = file;
-                }
-
-                generateTestFile(root, examples, componentFileMap);
-            });
-        } else {
-            generateTestFile(root, examples);
+    if (section.sections) {
```

Comment:
This is the new bit to support situations like how the modal docs use sections to specify examples.